File: Novomber 2022/Day 30/JAVA Code/Reorder List.java
// Problem Link :- https://practice.geeksforgeeks.org/problems/reorder-list/1
// Video Solution Link :- https://youtu.be/no3NIeoCamE

//{ Driver Code Starts
import java.util.*;

class Node {
    int data;
    Node next;
    Node(int d) {
        data = d;
        next = null;
    }
}

  public class ReorderList {
    Node head; // head of lisl
    Node last; // last of linked list

    /* Linked list Node*/

    /* Utility functions */

    /* Inserts a new Node at front of the list. */
    public void addToTheLast(Node node) {

        if (head == null) {
            head = node;
            last = head;
        } else {
            // Node temp = head;
            // while (temp.next != null) temp = temp.next;

            last.next = node;
            last = last.next;
        }
    }
    /* Function to print linked list */
    void printList() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    /* Drier program to test above functions */
    public static void main(String args[]) {

        /* Constructed Linked List is 1->2->3->4->5->6->
           7->8->8->9->null */
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0) {
            int n = sc.nextInt();
            ReorderList llist = new ReorderList();

            int a1 = sc.nextInt();
            Node head = new Node(a1);
            llist.addToTheLast(head);
            for (int i = 1; i < n; i++) {
                int a = sc.nextInt();
                llist.addToTheLast(new Node(a));
            }

            llist.head = new Solution().reorderlist(llist.head);

            llist.printList();

            t--;
        }
    }
}
// } Driver Code Ends


/* Following is the Linked list node structure */

/*class Node
{
    int data;
    Node next;
    Node(int d) {
        data = d;
        next = null;
    }
}*/

class Solution {
    // Time Complexity :- O(n);
    // Space Complexity :- O(1);
    
    Node reorderlist(Node head) {
        // Your code here
        int count=0;
        Node temp =head;
        while(temp!=null) {
            count++;
            temp=temp.next;
        }
        
        if(count==1 || count==2) return head;
        
        int left_count = count/2;
        if(count%2==1) left_count++;
        temp = head;
        while(left_count>0) {
            left_count--;
            temp = temp.next;
        }
        
        Node right = null;
        while(temp!=null) {
            Node node = new Node(temp.data);
            node.next=right;
            right=node;
            
            temp=temp.next;
        }
        
        Node ans=null;
        Node tail=null;
        while(right!=null) {
            Node node = new Node(head.data);
            if(ans==null) {
                ans=node;
                tail=node;
            } else {
                tail.next=node;
                tail=node;
            }
            head=head.next;
            
            node = new Node(right.data);
            tail.next=node;
            tail=node;
            right=right.next;
            
        }
        
        if(count%2==1) {
            Node node = new Node(head.data);
            tail.next=node;
        }
        return ans;
    }
}

File: Novomber 2022/Day 30/C++ Code/Reorder List.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/reorder-list/1
// Video Solution Link :- https://youtu.be/no3NIeoCamE

//{ Driver Code Starts
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <bits/stdc++.h>
using namespace std;
/* Linked list Node */
struct Node {
    int data;
    struct Node* next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};



// } Driver Code Ends
/* Following is the Linked list node structure */

/*
struct Node
{
    int data;
    struct Node* next;

    Node(int x){
        data = x;
        next = NULL;
    }
};
*/

class Solution{ 
public:
    // Time Complexity :- O(n);
    // Space Complexity :- O(n);
    
    void reorderList(Node* head) {
        // Your code here
        vector<Node*> list;
        while(head!=NULL) {
            list.push_back(head);
            head=head->next;
        }
        
        for(int i=0,j=list.size()-1;i<j;i++,j--) {
            if(list.size()%2==1 && i+2==j) {
                list[j]->next=list[i+1];
                list[i+1]->next=NULL;
            } else if(i+1!=j){
                list[j]->next=list[i+1];
            } else {
                list[j]->next=NULL;
            }
            
            list[i]->next=list[j];
        }
    }
};

//{ Driver Code Starts.



/* Function to create a new Node with given data */
struct Node* newNode(int data) {
    struct Node* new_Node = new Node(data);
    new_Node->data = data;
    new_Node->next = NULL;
    return new_Node;
}

void printList(struct Node* Node) {
    while (Node != NULL) {
        printf("%d ", Node->data);
        Node = Node->next;
    }
    printf("\n");
}

void freeList(struct Node* head) {
    struct Node* temp;
    while (head != NULL) {

        temp = head;
        head = head->next;
        free(temp);
    }
}

int main(void) {
    int t, n, m, i, x;

    cin >> t;
    while (t--) {
        struct Node* temp, *head;
        cin >> n;
        cin >> x;
        head = new Node(x);
        temp = head;
        for (i = 0; i < n - 1; i++) {
            cin >> x;
            temp->next = new Node(x);
            temp = temp->next;
        }
        
        Solution ob;
        
        ob.reorderList(head);
        printList(head);
        freeList(head);
    }
    return 0;
}

// } Driver Code Ends
